### PR TITLE
388: Add support for `musig` in descriptor templates

### PR DESCRIPTION
**Pull request by bigspider**
This required a few changes in the grammar describing descriptor templates: the single symbol `KP` is no longer enough to cleanly represent the syntax. Therefore, it's now replaced by a more fine-grained symbol hierarchy:
- `KEY`, mapping 1-on-1 to `KEY` expressions of descriptors
- `KP`, key placeholders, that represent a single root key, or an aggregate musig key; regardless of the nature of the key, key placeholders are what is followed by `/**` or `/<M;N>/*`.
- `KI`, key index, the actual "pointer" to the key information vector.

While the validity checks to avoid pubkey reuse before this change were feasible by looking at all the `@i` expressions in the descriptor template, they are now at the level of the KEY expressions, by comparing their key placeholders.
This means that the specs do not prevent having `@i` somewhere and `musig(@i,...)` elsewhere (which does not cause pubkey reuse in the actual scripts).

Note that there is no breaking change with the previous specs, just more fine-grained grammar in order to keep everything working in the presence of `musig`.

### CODE REVIEW

**Comment by bigspider** [comment]
cc @jgriffiths @benma

File: bip-0388.mediawiki [hunk 1]
```diff
@@ -147,25 +150,38 @@ See [[bip-0379.md|BIP-379]] for a precise specification of all the valid miniscr
 * An open brace <tt>{</tt>, a <tt>TREE</tt> expression, a comma <tt>,</tt>, a <tt>TREE</tt> expression, and a closing brace <tt>}</tt>
 
 
-<tt>KP</tt> expressions (key placeholders) consist of
-* a single character <tt>@</tt>
-* followed by a non-negative decimal number, with no leading zeros (except for <tt>@0</tt>)
+<tt>KEY</tt> expressions consist of
+* a <tt>KP</tt> expressions
```
Comment (jgriffiths) [inline bip-0388.mediawiki:154]: nit: expression, not expressions

**Comment by jgriffiths** [comment]
@bigspider thanks for the ping. Only one initial question before I review more thoroughly - why do we not allow derivation inside the musig expression KP elements, subject to the same rules as BIP-0390?

**Comment by bigspider** [comment]
> @bigspider thanks for the ping. Only one initial question before I review more thoroughly - why do we not allow derivation inside the musig expression KP elements, subject to the same rules as BIP-0390?

I did indeed suggest [removing that from descriptors as well](https://gnusha.org/pi/bitcoindev/CAMhCMoHLfRDZPeVcuDnDy4mXbsyOQbgbqWPK51VBxCW=a1wD-A@mail.gmail.com/T/), and I don't know of use cases that justify the (substantial) added complexity for signer.

It is much more inefficient in practice (especially for a transaction with many inputs coming from the same account, since the result of the key aggregation can be cached for all the inputs/change outputs, instead of repeating it each time with different keys.

Finally, it breaks the `key placeholder` concept, as it would imply that we need to support derivations at different nesting levels in the grammar.

File: bip-0388.mediawiki (hunk 1 above)
Comment (bigspider) [reply bip-0388.mediawiki]: Thanks! Fixed.

File: bip-0388.mediawiki [hunk 2]
```diff
@@ -1,3 +1,6 @@
+RECENT CHANGES:
+* (06 Nov 2024) Added <tt>musig</tt> key placeholders
```
[the remark this answers is not in the document]
Comment (bigspider) [reply bip-0388.mediawiki]: Indeed, I was trying to find examples of other BIPs, and I copied the style from [BIP-32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki), but a small 'Changelog' section like in [BIP-340](https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki#user-content-Changelog) is probably better. I'll do it later, once I don't expect further changes from the review.

File: bip-0388.mediawiki [hunk 3]
```diff
@@ -1,3 +1,6 @@
+RECENT CHANGES:
+* (06 Nov 2024) Added support for <tt>musig</tt> key placeholders in descriptor templates
+
```
Comment (murchandamus) [inline bip-0388.mediawiki:3]: The BIP formatting rules prescribe that the document starts with the preamble. May I suggest that you introduce a Change Log section in the [style of BIP 327](https://github.com/bitcoin/bips/blob/master/bip-0327.mediawiki#change-log) or [BIP 352](https://github.com/bitcoin/bips/blob/master/bip-0352.mediawiki#change-log) toward the end of the document where you incorporate this note?

Comment (jonatack) [reply bip-0388.mediawiki:3]: @murchandamus yes, see https://github.com/bitcoin/bips/pull/1697#discussion_r1834640167

@bigspider fwiw a suggested order at https://github.com/bitcoin/bips/pull/1691#discussion_r1817926795

**Comment by jgriffiths** [comment]
> It is much more inefficient in practice 

This is true but only applies if the feature is used. Support for it could be optional for example; wallets are free to limit the format and type of templates they support.

> Finally, it breaks the `key placeholder` concept, as it would imply that we need to support derivations at different nesting levels in the grammar.

IIUC this only requires that any wildcards in the musiq keys to aggregate must be derived at the same index as other keys when a wildcard is present, which is the same restriction as any wildcards in any key expression elsewhere. IOW, the entire expression must be derivable with a single value for all wildcards, whether inside musig() or elsewhere. Perhaps I'm failing to grok your meaning here, or perhaps things would be clearer if I worked through implementing support for this extension personally.

> I did indeed suggest [removing that from descriptors as well](https://gnusha.org/pi/bitcoindev/CAMhCMoHLfRDZPeVcuDnDy4mXbsyOQbgbqWPK51VBxCW=a1wD-A@mail.gmail.com/T/), and I don't know of use cases that justify the (substantial) added complexity for signers.

You raise reasonable points in the linked mail IMO. I can see that there may be a desire to not limit future use-cases, and I can also see that potentially allowing a mixture of fixed and derived keys to aggregate may have some uses (assuming its safe to do so). It is disappointing that (AFAICS) you did not receive a reply.

My concern here is that it appears on the descriptor side we have a technical decision with no clear rationale/documented justification, while on the descriptor template side there is an opposing technical decision with rationale but that may limit future interop or valuable use-cases.

I would prefer to have these two sides aligned, or at least the rationale for differing be explicit on both side (i.e. a reply to the points you raised on the record somewhere).

File: bip-0388.mediawiki (hunk 3 above)
Comment (bigspider) [reply bip-0388.mediawiki]: Good suggestion, done in eec10546357af967978a13aa0c989a31d4309b17.

**Comment by bigspider** [comment]
> This is true but only applies if the feature is used. Support for it could be optional for example; wallets are free to limit the format and type of templates they support.

If the language supports it, software wallet developers will use it; performance will probably be ok on their test transactions spending 1 UTXO; then one of their users who is trying to consolidate 50 UTXOs will discover it's incredibly slow, complain to the vendor, and there will be no way to fix it at that point.

To make a stronger case on the stark difference in performance: for a _n_-of-_n_ multisig and a transaction spending _t_ inputs
- aggregate-and-derive: 1 KeyAgg, plus 2 * _t_ BIP32 child pubkey derivations
- derive-and-aggregate: _t_ KeyAgg, plus 2 * _n_ * _t_ BIP32 child pubkey derivations
(some caching might be possible, but it doesn't change the substance). 

In other words, aggregate-and-derive doesn't cost significantly more than normal multisigs in hardware signer's performance once you have a few UTXOs (apart from needing 2 rounds, of course), while the penalty cost of derive-and-aggregate is enormous. BIP32 derivations are currently the biggest performance bottleneck in the Ledger bitcoin app, when using multi-user policies.

Therefore, by _not_ supporting it, I expect it will result in more efficient implementations, less debugging, less support tickets, and no loss of functionality.

> > Finally, it breaks the `key placeholder` concept, as it would imply that we need to support derivations at different nesting levels in the grammar.
> 
> IIUC this only requires that any wildcards in the musig keys to aggregate must be derived at the same index as other keys when a wildcard is present, which is the same restriction as any wildcards in any key expression elsewhere. IOW, the entire expression must be derivable with a single value for all wildcards, whether inside musig() or elsewhere. Perhaps I'm failing to grok your meaning here, or perhaps things would be clearer if I worked through implementing support for this extension personally.

Even just parsing a descriptor template supporting both versions becomes a lot harder. This is more or less what a key expression in my AST looks like:

```C
typedef struct {
    uint32_t num_first;   // NUM_a of /<NUM_a,NUM_b>/*
    uint32_t num_second;  // NUM_b of /<NUM_a,NUM_b>/*

    KeyExpressionType type;
    union {
        // type == 0
        struct {
            int16_t key_index;  // index of the key
        } k;
        // type == 1
        struct {
            rptr_musig_aggr_key_info_t musig_info;
        } m;
    };
} policy_node_keyexpr_t;
```

They are very simple objects, derivations can only go in one place, and the grammar to describe them is context-free (so parsing for them is based on a trivial lookahead on the next one or few characters).
 
> My concern here is that it appears on the descriptor side we have a technical decision with no clear rationale/documented justification, while on the descriptor template side there is an opposing technical decision with rationale but that may limit future interop or valuable use-cases.
> 
> I would prefer to have these two sides aligned, or at least the rationale for differing be explicit on both side (i.e. a reply to the points you raised on the record somewhere).

I understand the concern, and it is my desire to keep wallet policies as aligned as possible in order to minimize friction. 

However, wallet policies are not representing the same thing as descriptors:
- descriptors represent arbitrary outputs that can be spent with arbitrarily complicated spending logic that hardware signing devices might or might not support.
- wallet policies represent scripts for a vast class of smart contracts where the logic is however quite simple and uniform: "you're spending from an account, and change (if any) goes back to the same account".

Therefore, wallet policies do not need to be as general as descriptors. It is fine if they are opinionated and much more structured - in fact, this makes them more useful, by implicitly leading wallet developers to only use _sane_ descriptors.
It's the same reason why miniscript doesn't (and can't) support arbitrary Scripts, despite many other Scripts have perfectly valid use cases.

**Comment by jgriffiths** [comment]
@bigspider Thanks for the detailed response.

Note that as stated I don't disagree with your points on performance (original or as further fleshed out here), except to note that we already have an extension for non-bip44-style policies which e.g. Ledger doesn't support. However, on reflection I think there is nothing here that prevents derivation-before-aggregation being supported later, so this is not a blocker.

> BIP32 derivations are currently the biggest performance bottleneck in the Ledger bitcoin app, when using multi-user policies.

Somewhat OT, but you should implement the libwally optimization to avoid computing the parent fingerprint (a) when it is not needed at all (saves a has160 per derivation) and (b) when deriving a path (only the final derived key may need the fingerprint, the intermediate steps do not). In your impl, `get_derived_pubkey` never needs the fingerprint so adding a flag to `bip32_CKDpub` to skip calculating it would be a trivially implemented win there. In fact AFAICS none of your uses of that call require the fingerprint (e.g. PSBT).

> Even just parsing a descriptor template supporting both versions becomes a lot harder. 

Agreed the parsing gets more difficult. I will likely try adding support for this in wally to see how true that is for different impls.

> Therefore, wallet policies do not need to be as general as descriptors. It is fine if they are opinionated and much more structured - in fact, this makes them more useful, by implicitly leading wallet developers to only use _sane_ descriptors.

Disagree with the the _sane_ moniker but I accept/agree with this statement otherwise. I will final review and ack next week when I get a chance to compare implementing these changes with the PR contents, thanks.

**Comment by bigspider** [comment]
> Somewhat OT, but you should implement the libwally optimization to avoid computing the parent fingerprint (a) when it is not needed at all (saves a hash160 per derivation) and (b) when deriving a path (only the final derived key may need the fingerprint, the intermediate steps do not). In your impl, `get_derived_pubkey` never needs the fingerprint so adding a flag to `bip32_CKDpub` to skip calculating it would be a trivially implemented win there. In fact AFAICS none of your uses of that call require the fingerprint (e.g. PSBT).

I actually had this implemented few months ago but ended up not merging it because the performance impact is negligible for our implementation. It turns out that almost the entire cost of `bip32_CKDpub` is the single scalar multiplication. A one [line-change](https://github.com/LedgerHQ/app-bitcoin-new/commit/e066314100b6503f4e2d76e1d7c5f32357bc538f) with a faster scalar implementation ended up saving close to 50% total running time for some complex transaction! We still have some room for improvements there.

File: bip-0388.mediawiki [hunk 4]
```diff
@@ -80,7 +80,7 @@ We set two fundamental design goals:
 * Minimize the amount of information that is shown on screen - so that the user can actually validate it.
 * Minimize the number of times the user has to validate such information.
 
-Designing a secure protocol for the coordination of a descriptor wallet among distant parties is also a challenging problem that is out of scope in this document. See [[bip-0129.mediawiki|BIP-129 (Bitcoin Secure Multisig Setup)]] for an approach designed for multisignature wallets. Regardless of the approach, the ability for the user to carefully verify all the details of the spending policies using the hardware signer's screen is a prerequisite for security in adversarial environments.
+Designing a secure protocol for the coordination of a descriptor wallet among distant parties is also a challenging problem that is out of the scope of this document. See [[bip-0129.mediawiki|BIP-129 (Bitcoin Secure Multisig Setup)]] for an approach designed for multisig wallets. Regardless of the approach, the ability for the user to carefully verify all the details of the spending policies using the hardware signer's screen is a prerequisite for security in adversarial environments.
```
Comment (allysa08) [inline bip-0388.mediawiki:83]: - [ ]

**Review by murchandamus — COMMENTED** [review]
These changes appear to be only adding additional options, specifically to support MuSig2, and to improve a few phrases.

@bigspider: What would you consider the next steps? Are you waiting for some ACKs from stakeholders, or do you feel this is ready to go?

**Comment by bigspider** [comment]
> @bigspider: What would you consider the next steps? Are you waiting for some ACKs from stakeholders, or do you feel this is ready to go?

If nobody has further comments, it's good to go from my point of view. I already implemented MuSig2 support in the Ledger Bitcoin app matching these specs.

File: bip-0388.mediawiki [hunk 5]
```diff
@@ -147,25 +147,38 @@ See [[bip-0379.md|BIP-379]] for a precise specification of all the valid miniscr
...
+* followed by a non-negative decimal number, with no leading zeros (except for <tt>@0</tt>)
+
 The <tt>/**</tt> in the placeholder template represents commonly used paths for receive/change addresses, and is equivalent to <tt><0;1>/*</tt>.
 
 Note that while [[bip-0389.mediawiki|BIP-389]] allows multipath <tt>/<NUM;NUM;...;NUM></tt> expressions with an arbitrary number of options, this specification restricts it to exactly 2 choices (with the typical meaning of receive/change addresses).
 
-The placeholder <tt>@i</tt> for some number ''i'' represents the ''i''-th key in the vector of key information items (which must be of size at least ''i + 1'', or the wallet policy is invalid).
+<tt>SCRIPT</tt>, <tt>TREE</tt> and <tt>KEY</tt> expressions map directly to the corresponding concepts defined in [[bip-0380.mediawiki|BIP-380]] for output script descriptors.
+
+Each <tt>KEY</tt> expression always correspond to a precise public key in the final bitcoin Script. Therefore, all the derivation steps in the BIP-32 hierarchy are included in a <tt>KEY</tt> expression.
```
Comment (jonatack) [inline bip-0388.mediawiki:171]:
```suggestion
Each <tt>KEY</tt> expression always corresponds to a precise public key in the final bitcoin Script. Therefore, all the derivation steps in the BIP-32 hierarchy are included in a <tt>KEY</tt> expression.
```

File: bip-0388.mediawiki [hunk 6]
```diff
@@ -224,9 +239,10 @@ Common single-signature account patterns:
 * <tt>sh(wpkh(@0/**))</tt> (nested segwit).
 * <tt>tr(@0/**)</tt> (taproot single-signature account).
 
-Common multisignature schemes:
+Common multisig schemes:
 * <tt>wsh(multi(2,@0/**,@1/**))</tt> - SegWit 2-of-2 multisignature, keys in order.
```
Comment (jonatack) [inline bip-0388.mediawiki:243]: One last remaining use of "multisignature" if you wanted to convert them all to "multisig" in commit be5029f3224db8d5217a92a1ee7c90f000cf3d83

```suggestion
* <tt>wsh(multi(2,@0/**,@1/**))</tt> - SegWit 2-of-2 multisig, keys in order.
```

File: bip-0388.mediawiki [hunk 7]
```diff
@@ -151,7 +151,7 @@ See [[bip-0379.md|BIP-379]] for a precise specification of all the valid miniscr
 
 
 <tt>KEY</tt> expressions consist of
-* a <tt>KP</tt> expressions
+* a <tt>KP</tt> expression
```
Comment (jonatack) [inline bip-0388.mediawiki:151]: nit, would be clearer to make this change in the first commit a3928806b93bdf83

File: bip-0388.mediawiki [hunk 8]
```diff
@@ -311,7 +332,20 @@ Wallet policies are implemented in
...
 The reference implementation is for demonstration purposes only and not to be used in production environments.
 
-==Footnotes==
+== Change Log ==
+
+To help implementers understand updates to this document, we attach a version number that resembles ''semantic versioning'' (<code>MAJOR.MINOR.PATCH</code>).
+The <code>MAJOR</code> version is incremented if changes to the BIP are introduced that are incompatible with prior versions.
+An exception to this rule is <code>MAJOR</code> version zero (0.y.z) which is for development and does not need to be incremented if backwards incompatible changes are introduced.
+The <code>MINOR</code> version is incremented whenever the inputs or the output of an algorithm changes in a backward-compatible way or new backward-compatible functionality is added.
+The <code>PATCH</code> version is incremented for other changes that are noteworthy (bug fixes, test vectors, important clarifications, etc.).
```
Comment (jonatack) [inline bip-0388.mediawiki:341]: I think the paragraph describing a changelog can be omitted. We'll describe the changelogs in one place in BIP-2 (or any future replacement) and drop this paragraph from any other places.

File: bip-0388.mediawiki (hunk 7 above)
Comment (bigspider) [reply bip-0388.mediawiki:151]: Done

File: bip-0388.mediawiki (hunk 8 above)
Comment (bigspider) [reply bip-0388.mediawiki]: Nice, done!

**Review by murchandamus — APPROVED** [review]
LGTM

**Comment by jonatack** [comment]
Post-merge LGTM.

**Pull request merged**